### ISSUE TITLE
Create SilverStripe 5.x branch

### ISSUE THREAD

**Issue by robbieaverill** [silverstripe/silverstripe-frameworktest#69]
The SilverStripe installer now requires this module in Travis builds: https://github.com/silverstripe/silverstripe-installer/pull/283.

The master branch must be updated to support SilverStripe 5. We cannot merge up the installer branches until this is done.

**Comment by robbieaverill** [comment]
PR linked above. I've pushed the current master branch to `4` for SilverStripe 4 support.

**Comment by emteknetnz** [comment]
@robbieaverill 
Won't this break things?  I've seen a bunch of instances around the place with `composer require silverstripe/frameworktest:dev-master`

So now composer will get confused between installing framework 4 and 5?

**Comment by robbieaverill** [comment]
Probably. That’s why we don’t use dev-master. The solution is not doing that.

**Comment by robbieaverill** [comment]
I realise the irony in what I just said, so I’ll add a branch alias and update SS5 installer tomorrow

**Comment by robbieaverill** [comment]
> So now composer will get confused between installing framework 4 and 5?

No, it just won’t install frameworktest until you use the appropriate constraint. Existing lock files will still install fine, it’ll be mostly CI builds that are affected.

**Comment by dnsl48** [comment]
> I’ll add a branch alias and update SS5 installer tomorrow

Please don't make aliases :)
We just removed them from everywhere ([RFC 9344](https://github.com/silverstripe/silverstripe-framework/issues/9344)), and I think at this point it might be better to temporarily revert the change, until we're ready to go through all repositories and update the travis configs, so everything is consistent

**Comment by robbieaverill** [comment]
Do you mean revert the branching I did, or the introduction of a dependency in 4.x which is not able to be merged up to SilverStripe 5?

**Comment by robbieaverill** [comment]
As far as I can tell that linked RFC doesn't affect master branches. Despite the fact that it says "Remove all extra.branch-alias directives from every repository," framework, CMS, and installer still have them in the master branches.

In terms of this repository, what is done is all that is needed. If you want to push RFC 9344 to the extreme, then leaving this repo the way it is is correct, and we can continue to reference `frameworktest:dev-master` until the next time we do a major version release, and these references will all break again.

My recommendation would be to adjust your RFC retrospectively so it doesn't include master branches, add one here, then update the installer repository.

Either way, the master branch in this repository is now for SilverStripe 5, so this issue can be closed.

**Comment by dnsl48** [comment]
I was thinking about master of frameworktest requiring framework 5, rather than ^4 ([this PR](https://github.com/silverstripe/silverstripe-frameworktest/pull/70/files))

**Comment by dnsl48** [comment]
I'm not sure need to be changing the accepted RFCs. If we need to discuss something new, that could be a new RFC.
If the PR mentioned above will break a lot of the travis builds for the active supported branches, I'd say it's not worth it and we'll just revert it back.

**Comment by dnsl48** [comment]
> As far as I can tell that linked RFC doesn't affect master branches
The RFC defines: `1. Remove all extra.branch-alias directives from every repository.`

**Comment by robbieaverill** [comment]
Are you suggesting we change it to `dev-master`? If so, I'm really against that. Your RFC has half of the core team upvoting it, but mentions nothing about removing branch aliases from the master branches.

**Comment by robbieaverill** [comment]
Shall we continue this conversation on your accepted and implemented RFC?

**Comment by dnsl48** [comment]
> Are you suggesting we change it to dev-master? If so, I'm really against that.

I'm only suggesting we revert a self-merged PR that wasn't approved by anyone and makes a lot of pain for a bunch of maintainers.

**Comment by robbieaverill** [comment]
For the record, what I did is completely fine under the allowances for me being a SilverStripe core contributor.

I'm creating a new RFC to sort this out.

**Comment by dnsl48** [comment]
Fair enough. I'm not saying the PR is bad. I'm merely thinking it has some unforeseen side effects and might be worth reverting now.

**Comment by robbieaverill** [comment]
@dnsl48 I'd prefer you remove the dependency from the installer so we can still merge up the branches

**Comment by emteknetnz** [comment]
@robbieaverill I'm a little unclear what is meant by the 'remove the dependency from the installer' - is this what is being referred to? https://github.com/silverstripe/silverstripe-installer/pull/283/files#diff-354f30a63fb0907d4ad57269548329e3R104

If that's what is being referred to, I believe it was added because there was a new behat test in one of the modules that requires a special type of DataObject that is included in frameworktest.  Frameworktest was used to a provide 'test only' DataObject because `implements TestOnly` only works for phpunit tests not behat tests.

So I don't think that removing that requirement is really an option?

I may have misinterpreted what you're saying though, please correct me if I am mistaken

**Comment by dnsl48** [comment]
Ok, I suggest we step back and identify the problems we have with it.

I've mentioned the RFC because we've just removed the aliases, and I dislike the idea of adding them back, especially if that's ad-hoc fixes for some temperary issues. I think the RFC itself is not actually related to the problem we have in here, but I still think the branch aliases is not the best solution.

Here's what I see at the moment
 - frameworktest is not a direct dependency of installer, but rather is used for behat tests on Travis
 - frameworktest is not a core module, and its versions are not aligned with the core recipes (we currently have `0.1.0` and `dev-master` in use)
 - we _want_ to make this module consistent with others and follow the same semver principles
 - we're already half-way towards making it consistent

The options we have mentioned so far:
 - Option A: add a branch-alias
 - Option B: follow through the travis configs and update the frameworktest version used
 - Option C: revert the change to the frameworktest module (framework requirement)
 - Option D: revert the change to the installer travis (frameworktest requirement)

I don't like the Option A, because it introduces configuration inconsistencies redundant entities to maintain.
I don't like the Option D, because it would reduce the test coverage for the framework (although, we could remove it for 5 only)

My recommendation is we choosee between B and C

**Comment by robbieaverill** [comment]
@emteknetnz yes, that's what I'm referring to. I don't mind why you added it, but the point is that it's not compatible with SilverStripe 5.x, which means we (the maintainers) cannot merge up the branches any more. I've rectified that problem by creating a `4` branch in this repository, and making the master branch here compatible with SilverStripe 5. The 4.x installer branch will continue to work because it uses a tag, and we can now use `dev-master` in SilverStripe 5 CI builds for the framework test module.

You pointed out that it will become a problem again in SilverStripe 6, so I said I would add a branch alias for this repository so we can use `5.x-dev` instead of `dev-master`, thus reducing future maintenance burden.

@dnsl48:

>we want to make this module consistent with others and follow the same semver principles

No we don't. That's why we've not put branch aliases here in the past, nor released tags for it until recently. It's an internal module and will continue to be so I suspect.

You're missing option E, which is "do nothing" - I've already resolved this problem. I was going to add a branch alias so it doesn't become a problem again in future, but RFC 9344 prevents me from doing so (at the moment) I won't be doing that.

**Comment by dnsl48** [comment]
Haha, that's mean but fair, @robbieaverill :clap:

**Comment by emteknetnz** [comment]
@dnsl48 @robbieaverill 

Option D is not really a solution, this affects many more repos than just silverstripe/installer.  Also the silverstripe/installer uses ^0.1.0, not dev-master, so it shouldn't be affected by this PR.

Option A unfortunately breaks stuff right now without additional up-front work (detailed below)
(I'm assuming Option A is what was just merged, apologies if I have misinterpreted)

Option B would be certainly be nicer long term, though I'm not sure how much work this would be do to properly.  I'd be concerned particularly with older minor branches e.g. `4.4` we may need to do a security patches for that have an old `composer require silverstripe/frameworktest:dev-master` in there that will now be broken

It means we'd need to scan the `.travis.yml` file on every repo, several minor versions back.  I have recently done a similar thing manually though only for the latest minors, and it is a significant amount of manual labour that get pretty tiresome.

If we were to do this, I would certainly want to do some up-front work to create some automated tooling.  Inevitably we'd  increase the scope to "standarise all the travis things" while we're at it, which has been talked about a fair bit before and I believe had real long term value.  This would be a mini-project.  I'm not sure how this would fit in with all the other competing priorities we have though.  I think it will be a lot of work and not really something we can just sporadically do.

Option C revert the change to the frameworktest module, I think we might have to do this short term at least, right now this is what happens when I have `"frameworktest": "dev-master'` in my composer.json (this installed just fine a couple of days ago)
```
  Problem 1
    - silverstripe/framework 5.x-dev requires php >=7.2.0 -> your PHP version (7.3.7) overridden by "config.platform.php" version (7.1) does not satisfy that requirement.
    - silverstripe/frameworktest dev-master requires silverstripe/framework 5.x-dev -> satisfiable by silverstripe/framework[5.x-dev].
    - Installation request for silverstripe/frameworktest dev-master -> satisfiable by silverstripe/frameworktest[dev-master].
```

We cannot really have this in our travis builds, it means people will have random failures going forward on current branches as well as older minors.  That is really not good for our workflow.  We're mid release cycle right now so it's not good timing

What might be good is if we now:
a) Revert this change for now for short-term pragmatic reasons (the change breaks travis builds which we rely on)
b) Raise a new issue to scope out the work required to reach a good long-term outcome which satisfies everyones particular requirements

How does that sound?

**Comment by robbieaverill** [comment]
> @dnsl48 Haha, that's mean but fair, @robbieaverill 👏

I do not intend to be mean, so I'm sorry if that's how my comments have come across.

---

>We cannot really have this in our travis builds, it means people will have random failures going forward on current branches as well as older minors

>I think we might have to do this short term at least, right now this is what happens when I have "silverstripe/recipe-cms": "4.x-dev" + "frameworktest": "dev-master' in my composer.json (this installed just fine a couple of days ago)

>That is really not good for our workflow. We're mid release cycle right now so it's not good timing

These are all great reasons to use branch aliases instead of `dev-master`, and to not introduce dependencies into the core repositories that are not compatible with the release lines they'll be merged up into.

@emteknetnz I don't want you to revert anything, unless it's your introduction of this module into the core CI builds.

I'm more than happy to fix the builds as they break if you'd like. I'll go through them tonight and fix the ones I see that are referencing `frameworktest:dev-master` and change to `4.x-dev`. I will **not** change those constraints for SilverStripe 5 branches, since we currently have an RFC that prevents me from doing that.

**Comment by dnsl48** [comment]
Sounds like a plan :+1: 
Let's move on and just fix the builds. To be fair, I think in reality there won't be too many of them. I will also volunteer to fix the red ones when I see it.
I don't think it will interfere with the upcoming releases in any way, as we're on RC1 stage now, so no worries about that.

Thank you for your work @robbieaverill!

**Comment by emteknetnz** [comment]
> the point is that it's not compatible with SilverStripe 5.x, which means we (the maintainers) cannot merge up the branches any more
I can see how this would be hugely frustrating for you

I think everyone agrees that having travis config files with `frameworktest: dev-master` is not good and `frameworktest: 4.x-dev` is much better

@robbieaverill if you're happy to do some of the leg work yourself to fix the travis config files then that's certainly appreciated

**Comment by robbieaverill** [comment]
>I think everyone agrees that having travis config files with frameworktest: dev-master is not good and frameworktest: 4.x-dev is much better

Unfortunately that's not true @emteknetnz, as our current stance says "we don't do branch aliases", therefore we still have config files with that constraint in them for SilverStripe 5. Let's continue that conversation on https://github.com/silverstripe/silverstripe-framework/issues/9560 though.

**Comment by emteknetnz** [comment]
I don't understand .. we now have a `4` branch of `frameworktest` so we can just use `4-x.dev` now?  That is not a branch-alias, it's an actual branch?

That is something I like.  For me the downside is the broken travis builds and toil involved in fixing them.

**Comment by robbieaverill** [comment]
Yes, 4.x is fine, but 5.x is not. That is the reason I opened this issue, I share your concerns.
